### ISSUE TITLE
PrettyMarkupFormatter loses NoBreakSpaces/&nbsp; at the beginning or end of the content

### ISSUE THREAD

**Issue by Tetedeiench** [AngleSharp/AngleSharp#869]
# Bug Report

AngleSharp 0.14.0

## Description

Tidying an HTML document removes NoBreakSpaces ( nbsp ) characters from the output if they are located at the beginning or the end of a line (or even unique)

## Steps to Reproduce

Given the following example :
```C#
        static void Main(string[] args)
        {
            var documentStr = "<html><head></head><body><p>&nbsp;</p></body></html>";
            var parser = new HtmlParser();
            
            var document = parser.ParseDocument(documentStr);
            var sw = new StringWriter();

            document.ToHtml(sw, new PrettyMarkupFormatter());

            Console.WriteLine(sw.ToString());
            Console.ReadLine();
        }
```
**Expected behavior:** 

```
<html>
        <head></head>
        <body>
                <p>&nbsp;</p>
        </body>
</html>
```

**Actual behavior:**
```
<html>
        <head></head>
        <body>
                <p></p>
        </body>
</html>
```

**Environment details:** Windows 10, .Net Framework 4.6.1

## Possible Solution

To me (i'm no expert), in PrettyMarkupFormatter.Text, you're using TrimEnd() and TrimStart(), which will remove any space characters, inclusing 0xa0 (representing &nbsp; ). 

I think they should be kept, as removing those can break the display ( <p></p> and <p>&nbsp;</p> are different in HTML/CSS and browsers render them differently if they are empty).

This should at least be configurable through an option somewhere (it actually broke the display of my HTML document by removing those).

**Comment by FlorianRappl** [comment]
> HTML/CSS and browsers render them differently if they are empty).

Actually its a bit more complicated than that. The observation is certainly right that *any* whitespace character *may* result in a rendered whitespace, however, there are certain conditions (e.g., CSS `display`, parsing mode, ...) that need to be met, too.

The pretty markup formatter was never thought to be rendering preserving. It *cannot* be. Why? Because *inserting* whitespace has potentially the same effect as *removing* whitespace, i.e., we change the rendering.

Consider

```html
<p><b>Hello</b><i>there</i><u>my</u><span>friend</span></p>
```

pretty printed we would change to have

```html
<p>
  <b>Hello</b>
  <i>there</i>
  <u>my</u>
  <span>friend</span>
</p>
```

Suddenly we have spaces in between.

So long story short, I'm not sure if we should / could treat this. If you want to preserve the rendering then the pretty printer is not for you. It will also never be, because all we can / could do is to change single whitespace to be multiple whitespaces (or vice-versa).

**Comment by FlorianRappl** [comment]
I'll leave this one open for the moment (for a potential discussion esp. around potential options for the `PrettyMarkupFormatter`).

**Comment by Tetedeiench** [comment]
Hi Florian, thanks for your answer !

Thinking about it, the issue isn't about empty spaces solely, It's transforming tags with a content (might it be spaces, nbsp, etc) into tags without content. This can have a great influence on your actual code. I want it to be pretty to read so it is easy to modify, but i don't want it to break because i pretty-printed it ;)

For the following, please remember i'm not an expert in HTML/CSS by far.

Consider the following (terrible) example i put together real quick. This isn't pretty code or something i'd use in a real scenario, it's just for the sake of demonstration : 
https://jsfiddle.net/6842cx9f/

The text is centered next to the red square.

Remove the contents in the red square p tag : 
https://jsfiddle.net/6842cx9f/1/

Suddenly the text isn't centered anymore.

While i am expecting a pretty printer to modify indentation, whitespaces, and so on, i'm not expecting it to transform a tag with contents to a tag without contents.

**Comment by FlorianRappl** [comment]
> While i am expecting a pretty printer to modify indentation, whitespaces, and so on, i'm not expecting it to transform a tag with contents to a tag without contents.

As outlined this is what the pretty printer can / will do. Otherwise it would not be able to pretty print, because the inserted HTML needs to remain (minus expansion / collapsing of existing whitespace). I guess I explained that already.

If you want to preserve your HTML then stay with the standard formatter or use the minification formatter to reduce overhead.

**Remark**: There is no need to put together any example, because I already showed you that the pretty printer is not preserving (and it was never thought to be -> see explanation). The only reason to keep this one open is *to discuss potential options* (if there are any).

**Comment by Tetedeiench** [comment]
Alright ! 

As for potential options, i'd personally include one as "DontEmptyTags" or something along those lines.

Going further in discussing whether it is a good idea or not is definitely out of my league, so if that's the expected behaviour, i'd still describe it somewhere, and close this issue.

Thanks !

**Comment by TopCoder02** [comment]
@Tetedeiench  I had the same issue, this is what I did for a workaround.  Essentially, there are two ways a Non-Breaking Space can be represented, one is an ASCII character that represents it but looks like a normal space, the other is the way you described. 

@FlorianRappl I think an option like this can be added to the formatter to handle the scenario @Tetedeiench describe and you can use the code below, or something similar to accomplish it.  

Three years late, but I'm sure this will help others who have the same issue. 

   ```
 Private Function MakePrettyHTML(ByVal vHTMLString As String) As String
        Dim PrettyParser As New AngleSharp.Html.Parser.HtmlParser
        vHTMLString = Regex.Replace(vHTMLString, "(?<=>) (?=[ \r\n\t<\/])", "~NBS~", RegexOptions.IgnoreCase Or RegexOptions.IgnorePatternWhitespace)
        vHTMLString = Regex.Replace(vHTMLString, "(?<=>)&nbsp; (?=[ \r\n\t<\/])", "~NBS2~", RegexOptions.IgnoreCase Or RegexOptions.IgnorePatternWhitespace)
        Dim oDocument = PrettyParser.ParseDocument(vHTMLString)
        Dim sw = New StringWriter()
        oDocument.ToHtml(sw, New PrettyMarkupFormatter())
        Dim oPrettyHTML As String = sw.ToString()
        oPrettyHTML = oPrettyHTML.Replace("~NBS~", " ")
        oPrettyHTML = oPrettyHTML.Replace("~NBS2~", "&nbsp;")
    End Function
```
